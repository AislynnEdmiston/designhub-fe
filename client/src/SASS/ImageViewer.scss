// .ImageViewer {
//   // border: 1px solid white;
//   color: white;
//   display: flex;
//   height: 68vh;
//   width: 100%;

//   main.ImageViewer__body {
//     // border: 1px solid white;
//     align-items: flex-start;
//     display: flex;
//     flex-flow: row nowrap;
//     justify-content: space-between;
//     margin: 0;
//     width: 100%;

//     .ImageViewer__main-image {
//       background-color: #15161a;
//       border: 5px solid #15161a;
//       box-sizing: border-box;
//       border-radius: 10px;
//       cursor: pointer;
//       overflow: hidden;
//       text-align: center;
//       height: 68vh;
//       width: 100%;
//       display: flex;
//       align-items: center;
//       justify-content: center;

//       img {
//         max-width: 100%;
//         max-height: 100%;
//       }
//     }

//     .ImageViewer__thumbnails {
//       display: flex;
//       flex-flow: column nowrap;
//       width: 50%;
//       height: 68vh;
//       overflow-y: scroll;
//       justify-content: flex-start;
//       align-items: center;

//       img {
//         border: 5px solid #15161a;
//         box-sizing: border-box;
//         border-radius: 10px;
//         margin-bottom: 16px;
//         height: 151px;
//         width: 211px;
//         object-fit: cover;

//         &:hover {
//           border: 5px solid #5557fe;
//           cursor: pointer;
//         }
//       }
//     }

//     .modal--close {
//       display: none;
//     }

//     .modal--expand {
//       animation: fadeIn 0.2s ease-in-out;
//       display: flex;
//       align-items: center;
//       justify-content: center;
//       text-align: center;
//       opacity: 1;
//       overflow-y: scroll;
//       position: absolute;
//       top: 0;
//       left: 0;
//       z-index: 999999;
//       width: 100%;
//       height: 100%;

//       img {
//         position: relative;
//         max-width: 80%;
//         z-index: 999;
//       }

//       span {
//         background: rgba(0, 0, 0, 0.897);
//         cursor: pointer;
//         content: '';
//         position: fixed;
//         left: 0;
//         top: 0;
//         width: 100%;
//         height: 100vh;
//         z-index: 99;
//       }

//       @keyframes fadeIn {
//         from {
//           opacity: 0;
//         }
//         to {
//           opacity: 1;
//         }
//       }
//     }
//   }
// }

// ::-webkit-scrollbar {
//   width: 0px; /* Remove scrollbar space */
//   background: transparent; /* Optional: just make scrollbar invisible */
// }
