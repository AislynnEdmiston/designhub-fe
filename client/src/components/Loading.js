import React from 'react';
import bodymovin from 'bodymovin';

import loading from '../ASSETS/loading.json';

import '../SASS/Loading.scss';

const Loading = () => {
  var svgContainer = document.getElementById('svgContainer');
  var animItem = bodymovin.loadAnimation({
    wrapper: svgContainer,
    animType: 'svg',
    loop: true,
    path: 'https://assets5.lottiefiles.com/datafiles/AX0rqrGV5ahKpWr/data.json'
  });
  return (
    <div className="Loading-container">
      <div id="svgContainer"></div>
      <p className="loading-text">Loading awesome content...</p>
    </div>
  );
};

export default Loading;
