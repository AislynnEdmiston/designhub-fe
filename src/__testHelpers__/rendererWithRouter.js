import React from 'react'
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux'
import { createStore, applyMiddleware } from 'redux'
import { createMemoryHistory } from 'history';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';

import rootReducer from '../store/reducers';
jest.mock('../auth-wrapper');

const store = createStore(
  rootReducer,
  applyMiddleware(thunk)
);
const rendererWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) => {
  return {
    ...renderer.create(
      <Provider store={store}>
        <Router history={history}>
          {ui}
        </Router>
      </Provider>
    ),
    history
  }
}

export default rendererWithRouter;
