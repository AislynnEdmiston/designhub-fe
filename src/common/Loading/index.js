import React from 'react';
import './styles.scss';

export default function Loading() {
  return (
    <div className="loading-wrapper">
      <h1>Loading...</h1>
    </div>
  );
}
