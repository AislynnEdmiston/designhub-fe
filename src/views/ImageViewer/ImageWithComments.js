import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import { axiosWithAuth } from '../../utilities/axiosWithAuth.js';
import { connect } from 'react-redux';

import { StickyComment } from './StickyComment';
import { TempComment } from './TempComment';
import ModalXIcon from '../../common/Icons/ModalXIcon.js';
import CommentBubbleIcon from '../../common/Icons/CommentBubbleIcon.js';
import './SASS/StickyComment.scss';

import {
  addPhotoComment,
  getPhotoComments,
  getProjectComments
} from '../../store/actions';
const uuidv1 = require('uuid/v1');

class ImageWithComments extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activeImg: this.props.activeImg,
      comments: [],
      hidden: false,
      tempComments: []
    };
    //location of the dom node, this is definitely a better way to do this using refs
    this.element = null;
    this.leftOffset = null;
    this.topOffset = null;
  }

  render() {
    const { id, url } = this.props.activeImg;
    const { hidden, tempComments } = this.state;
    //activeComments are EXISTING COMMENTS related to the current image
    let activeComments = this.props.photoComments;
    // tempComments are the sticky comment submit form
    let activeTemp = tempComments.filter(c => c.imageId === id);
    return (
      <>
        <div className="StickyComments__TopBar">
          <p>Click anywhere on the image to leave a sticky comment</p>
          <section className="StickyComments__TopBar__Icons">
            <div
              className={hidden ? 'hidden' : null}
              onClick={() => this.setState({ hidden: !this.state.hidden })}
            >
              <CommentBubbleIcon />
            </div>
            <div onClick={this.props.closeModal}>
              <ModalXIcon className="close" />
            </div>
          </section>
        </div>

        <div className="ImageWithComments">
          <img
            alt={url}
            className="ImageWithComments__full-img"
            onClick={e => this.handleClick(e, id)}
            src={url}
          />

          {activeTemp[0] &&
            activeTemp.map(c => (
              <TempComment
                c={c}
                key={c.id}
                hidden={this.state.hidden}
                onSubmit={this.handleSubmit}
                commentDelete={this.commentDelete}
              />
            ))}
          {activeComments[0] &&
            activeComments.map(s => (
              <StickyComment
                {...s}
                hidden={this.state.hidden}
                key={s.id}
                commentDelete={this.commentDelete}
              />
            ))}
        </div>
      </>
    );
  }
  componentDidMount() {
    //refactor this into refs?
    this.element = findDOMNode(this);
    // GET ALL COMMENTS
    const projectComments = this.props.comments;
    this.props.getPhotoComments(this.props.activeImg.id);
    this.setState({
      ...this.state,
      comments: projectComments
    });
  }

  makeComment = (id, x, y) => {
    const newComment = {
      id: uuidv1(),
      imageId: id,
      userId: this.props.activeUser.id,
      username: this.props.activeUser.username,
      projectId: this.props.thisProject.id,
      editing: false,
      // left&top are coordinates used to render the sticky comment dot, they are in relation to the dom element (this.element)
      left: `${x}px`,
      top: `${y}px`,
      text: ''
    };
    this.setState({
      ...this.state,
      tempComments: [newComment]
    });
  };

  handleClick = async (e, id) => {
    e.persist();
    //grab the mouseclick location and apply offset from the parent dom node, there is probably a better way to do this
    let x = e.nativeEvent.clientX;
    let y = e.nativeEvent.clientY;
    await this.updateElementPosition(x, y);
    x = this.leftOffset;
    y = this.topOffset;
    this.makeComment(id, x, y);
  };

  postCommentNotification = async (
    username,
    commentText,
    projectId,
    invitedUserId,
    activeUserId,
    mainImgUrl,
    commentsId,
    activeUserAvatar,
    type
  ) => {
    axiosWithAuth().post('api/v1/invite/comments', {
      activeUsername: username,
      commentText: commentText,
      projectId: projectId,
      invitedUserId: invitedUserId,
      activeUserId: activeUserId,
      mainImgUrl: mainImgUrl,
      commentsId: commentsId,
      activeUserAvatar: activeUserAvatar,
      type: type
    });
  };

  handleSubmit = async (e, c) => {
    e.preventDefault();
    const { comments } = this.state;
    const thisComment = { ...c };

    // delete local state flags before submitting to database
    delete thisComment.id;
    delete thisComment.editing;

    try {
      const res = await this.props.addPhotoComment(thisComment);
      const newComment = res.data.data[0];
      const updateComments = [...comments, newComment];
      this.props.getPhotoComments(this.props.activeImg.id);
      this.props.getProjectComments(this.props.thisProject.id);
      this.setState({
        ...this.state,
        tempComments: [],
        comments: updateComments
      });

      if (this.props.activeUser.id !== this.props.thisProject.userId) {
        await this.postCommentNotification(
          this.props.activeUser.username,
          newComment.text,
          this.props.thisProject.id,
          this.props.thisProject.userId,
          this.props.activeUser.id,
          this.props.activeImg.url,
          newComment.id,
          this.props.activeUser.avatar,
          'comment'
        );
      }
    } catch (err) {
      console.log('ImageWithComments.js handleSubmit() ERROR', err);
    }
  };

  updateElementPosition = (x, y) => {
    //get the size of the dom element
    const rect = this.element.getBoundingClientRect();
    // compensate for window scroll position
    this.topOffset = rect.top > 0 ? y : y - rect.top;
    this.leftOffset = rect.left > 0 ? x : x - rect.left;
  };
}

const mapStateToProps = state => {
  return {
    photoComments: state.comments.photoComments
  };
};

export default connect(
  mapStateToProps,
  { addPhotoComment, getPhotoComments, getProjectComments }
)(ImageWithComments);
