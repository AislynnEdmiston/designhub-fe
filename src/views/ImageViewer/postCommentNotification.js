import { axiosWithAuth } from '../../utilities/axiosWithAuth.js';

const postCommentNotification = async (
  username,
  commentText,
  projectId,
  invitedUserId,
  activeUserId,
  mainImgUrl,
  commentsId,
  activeUserAvatar,
  type
) => {
  axiosWithAuth().post('api/v1/invite/comments', {
    activeUsername: username,
    commentText: commentText,
    projectId: projectId,
    invitedUserId: invitedUserId,
    activeUserId: activeUserId,
    mainImgUrl: mainImgUrl,
    commentsId: commentsId,
    activeUserAvatar: activeUserAvatar,
    type: type,
  });
};

export default postCommentNotification;
