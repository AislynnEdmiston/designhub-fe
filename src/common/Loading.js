import React from 'react';
import { GooSpinner } from 'react-spinners-kit';
import './SASS/Loading.scss';

const Loading = () => {
  return (
    <div className="Loading-container">
      <GooSpinner size={80} color="#5557fe" className="loader" />
    </div>
  );
};

export default Loading;
