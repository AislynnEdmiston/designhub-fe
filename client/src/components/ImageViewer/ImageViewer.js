import React, { Component } from 'react';

import ImageWithComments from './ImageWithComments';
import ProjectComments from './ProjectComments.js';
import defaultImage from '../../ASSETS/default_thumbnail.svg';
import Loading from '../Loading';

import '../../SASS/ImageViewer.scss';

class ImageViewer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activeImg: null,
      comments: [],
      modal: false
    };
  }

  componentDidMount() {
    this.setState({
      ...this.state,
      comments: this.props.comments
    });
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.thumbnails !== prevProps.thumbnails) {
      this.setState({
        activeImg: this.props.thumbnails[0]
      });
    }
  }

  changeImg = imgObj => {
    if (
      this.state.activeImg === null ||
      this.state.activeImg.id !== imgObj.id
    ) {
      this.setState({ activeImg: imgObj });
    }
  };

  closeModal = () => {
    this.setState({ modal: false });
  };

  render() {
    const { activeImg, modal } = this.state;

    if (activeImg === null) {
      return <Loading />;
    } else {
      return (
        <>
          <div className="ImageViewer">
            <main className="ImageViewer__body">
              <div className={modal ? 'modal--expand' : 'modal--close'}>
                <span
                  className="modal--expand__background-overlay"
                  onClick={this.closeModal}
                />
                {modal && (
                  <ImageWithComments
                    activeUser={this.props.activeUser}
                    activeImg={activeImg}
                    closeModal={this.closeModal}
                    comments={
                      this.state.comments.length > 0
                        ? this.state.comments
                        : this.props.comments
                    }
                    thisProject={this.props.thisProject}
                  />
                )}
              </div>
              <div className="main-image-container">
                <section className="ImageViewer__main-image">
                  {!activeImg ? (
                    <img
                      style={{ cursor: 'default' }}
                      src={defaultImage}
                      alt="main project"
                      className="main-image"
                    />
                  ) : (
                    <img
                      src={
                        activeImg ? activeImg.url : this.props.thumbnails[0].url
                      }
                      alt="main project"
                      onClick={() => this.setState({ modal: true })}
                      className="main-image"
                    />
                  )}
                </section>
                <section className="ImageViewer__thumbnails">
                  {this.props.thumbnails.map(t => (
                    <img
                      key={t.url}
                      src={t.url}
                      alt="project-thumbnail"
                      onClick={() => this.changeImg(t)}
                      className="thumbnails"
                    />
                  ))}
                </section>
              </div>
            </main>
          </div>
          <ProjectComments
            activeUser={this.props.activeUser}
            modal={modal}
            thisProject={this.props.thisProject}
          />
        </>
      );
    }
  }
}

export default ImageViewer;
