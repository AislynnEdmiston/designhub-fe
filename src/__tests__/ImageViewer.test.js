import React from 'react';
import { shallow } from 'enzyme';

import ImageViewer from '../views/ImageViewer/ImageViewer';
import rendererWithRouter from '../__testHelpers__/rendererWithRouter';
import Loading from '../common/Loading';
describe('<ImageViewer />', () => {
  describe('should match the previous snapshot of the component', () => {
    it('tests that the component has not changed and matched the previous snapshot', () => {
      const activeImg ={
        id: 1
      }
      const tree = rendererWithRouter(
        <ImageViewer activeImg={activeImg} history={{ location: { pathname: '/' } }} />
      );
      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe('<ImageViewer/>', () => {
    const wrapper = shallow(<ImageViewer/>);

    it('should render div element successfuly', () => {
    expect(wrapper.find(ImageViewer).render().find('.ImageViewer')).to.have.lengthOf(1);
    });
  });
});
