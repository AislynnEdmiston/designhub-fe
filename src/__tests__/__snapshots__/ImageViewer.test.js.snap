// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageViewer /> should match the previous snapshot of the component tests that the component has not changed and matched the previous snapshot 1`] = `
<div
  className="Loading-container"
>
  <div
    className="goo__Wrapper-sc-12keask-0 crDyBj"
    size={80}
  >
    <div
      className="goo__BallsWrapper-sc-12keask-1 fvLLXo"
      size={80}
    >
      <div
        className="goo__Ball-sc-12keask-2 fviCBF"
        color="#5557fe"
        size={80}
        x={26.666666666666664}
        y={26.666666666666664}
      />
      <div
        className="goo__Ball-sc-12keask-2 dctINf"
        color="#5557fe"
        size={80}
        x={26.666666666666664}
        y={26.666666666666664}
      />
    </div>
    <svg
      version="1.1"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <filter
          id="goo"
        >
          <feGaussianBlur
            in="SourceGraphic"
            result="blur"
            stdDeviation="6"
          />
          <feColorMatrix
            in="blur"
            mode="matrix"
            result="goo"
            values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 15 -5"
          />
          <feBlend
            in="SourceGraphic"
            in2="goo"
          />
        </filter>
      </defs>
    </svg>
  </div>
</div>
`;
